File: qt/gen_qinputdevice.cpp
// +build ignore

#include <QInputDevice>
#include <QList>
#include <QMetaObject>
#include <QObject>
#include <QRect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qinputdevice.h>
#include "gen_qinputdevice.h"

class MiqtVirtualQInputDevice : public virtual QInputDevice {
public:

	MiqtVirtualQInputDevice(): QInputDevice() {};
	MiqtVirtualQInputDevice(const QString& name, qint64 systemId, DeviceType typeVal): QInputDevice(name, systemId, typeVal) {};
	MiqtVirtualQInputDevice(QObject* parent): QInputDevice(parent) {};
	MiqtVirtualQInputDevice(const QString& name, qint64 systemId, DeviceType typeVal, const QString& seatName): QInputDevice(name, systemId, typeVal, seatName) {};
	MiqtVirtualQInputDevice(const QString& name, qint64 systemId, DeviceType typeVal, const QString& seatName, QObject* parent): QInputDevice(name, systemId, typeVal, seatName, parent) {};

	virtual ~MiqtVirtualQInputDevice() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QInputDevice::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QInputDevice_MetaObject(const_cast<MiqtVirtualQInputDevice*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QInputDevice::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QInputDevice::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QInputDevice_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QInputDevice::qt_metacast(param1);

	}

};

QInputDevice* QInputDevice_new() {
	return new MiqtVirtualQInputDevice();
}

QInputDevice* QInputDevice_new2(struct miqt_string name, long long systemId, DeviceType typeVal) {
	QString name_QString = QString::fromUtf8(name.data, name.len);
	return new MiqtVirtualQInputDevice(name_QString, static_cast<qint64>(systemId), typeVal);
}

QInputDevice* QInputDevice_new3(QObject* parent) {
	return new MiqtVirtualQInputDevice(parent);
}

QInputDevice* QInputDevice_new4(struct miqt_string name, long long systemId, DeviceType typeVal, struct miqt_string seatName) {
	QString name_QString = QString::fromUtf8(name.data, name.len);
	QString seatName_QString = QString::fromUtf8(seatName.data, seatName.len);
	return new MiqtVirtualQInputDevice(name_QString, static_cast<qint64>(systemId), typeVal, seatName_QString);
}

QInputDevice* QInputDevice_new5(struct miqt_string name, long long systemId, DeviceType typeVal, struct miqt_string seatName, QObject* parent) {
	QString name_QString = QString::fromUtf8(name.data, name.len);
	QString seatName_QString = QString::fromUtf8(seatName.data, seatName.len);
	return new MiqtVirtualQInputDevice(name_QString, static_cast<qint64>(systemId), typeVal, seatName_QString, parent);
}

void QInputDevice_virtbase(QInputDevice* src, QObject** outptr_QObject) {
	*outptr_QObject = static_cast<QObject*>(src);
}

QMetaObject* QInputDevice_MetaObject(const QInputDevice* self) {
	return (QMetaObject*) self->metaObject();
}

void* QInputDevice_Metacast(QInputDevice* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QInputDevice_Tr(const char* s) {
	QString _ret = QInputDevice::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QInputDevice_Name(const QInputDevice* self) {
	QString _ret = self->name();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

DeviceType QInputDevice_Type(const QInputDevice* self) {
	return self->type();
}

Capabilities QInputDevice_Capabilities(const QInputDevice* self) {
	return self->capabilities();
}

bool QInputDevice_HasCapability(const QInputDevice* self, Capability cap) {
	return self->hasCapability(cap);
}

long long QInputDevice_SystemId(const QInputDevice* self) {
	qint64 _ret = self->systemId();
	return static_cast<long long>(_ret);
}

struct miqt_string QInputDevice_SeatName(const QInputDevice* self) {
	QString _ret = self->seatName();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

QRect* QInputDevice_AvailableVirtualGeometry(const QInputDevice* self) {
	return new QRect(self->availableVirtualGeometry());
}

struct miqt_array /* of struct miqt_string */  QInputDevice_SeatNames() {
	QStringList _ret = QInputDevice::seatNames();
	// Convert QList<> from C++ memory to manually-managed C memory
	struct miqt_string* _arr = static_cast<struct miqt_string*>(malloc(sizeof(struct miqt_string) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		QString _lv_ret = _ret[i];
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray _lv_b = _lv_ret.toUtf8();
		struct miqt_string _lv_ms;
		_lv_ms.len = _lv_b.length();
		_lv_ms.data = static_cast<char*>(malloc(_lv_ms.len));
		memcpy(_lv_ms.data, _lv_b.data(), _lv_ms.len);
		_arr[i] = _lv_ms;
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

struct miqt_array /* of QInputDevice* */  QInputDevice_Devices() {
	QList<const QInputDevice *> _ret = QInputDevice::devices();
	// Convert QList<> from C++ memory to manually-managed C memory
	QInputDevice** _arr = static_cast<QInputDevice**>(malloc(sizeof(QInputDevice*) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = (QInputDevice*) _ret[i];
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

QInputDevice* QInputDevice_PrimaryKeyboard() {
	return (QInputDevice*) QInputDevice::primaryKeyboard();
}

bool QInputDevice_OperatorEqual(const QInputDevice* self, QInputDevice* other) {
	return (*self == *other);
}

void QInputDevice_AvailableVirtualGeometryChanged(QInputDevice* self, QRect* area) {
	self->availableVirtualGeometryChanged(*area);
}

void QInputDevice_connect_AvailableVirtualGeometryChanged(QInputDevice* self, intptr_t slot) {
	MiqtVirtualQInputDevice::connect(self, static_cast<void (QInputDevice::*)(QRect)>(&QInputDevice::availableVirtualGeometryChanged), self, [=](QRect area) {
		QRect* sigval1 = new QRect(area);
		miqt_exec_callback_QInputDevice_AvailableVirtualGeometryChanged(slot, sigval1);
	});
}

struct miqt_string QInputDevice_Tr2(const char* s, const char* c) {
	QString _ret = QInputDevice::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QInputDevice_Tr3(const char* s, const char* c, int n) {
	QString _ret = QInputDevice::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

QInputDevice* QInputDevice_PrimaryKeyboard1(struct miqt_string seatName) {
	QString seatName_QString = QString::fromUtf8(seatName.data, seatName.len);
	return (QInputDevice*) QInputDevice::primaryKeyboard(seatName_QString);
}

void QInputDevice_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQInputDevice*>( (QInputDevice*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QInputDevice_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQInputDevice*)(self) )->virtualbase_MetaObject();
}

void QInputDevice_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQInputDevice*>( (QInputDevice*)(self) )->handle__Metacast = slot;
}

void* QInputDevice_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQInputDevice*)(self) )->virtualbase_Metacast(param1);
}

void QInputDevice_Delete(QInputDevice* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQInputDevice*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qinputdevice.go
package qt

import (
	"unsafe"
)

type QInputDevice__DeviceType int

const (
	QInputDevice__Unknown     QInputDevice__DeviceType = 0
	QInputDevice__Mouse       QInputDevice__DeviceType = 1
	QInputDevice__TouchScreen QInputDevice__DeviceType = 2
	QInputDevice__TouchPad    QInputDevice__DeviceType = 4
	QInputDevice__Puck        QInputDevice__DeviceType = 8
	QInputDevice__Stylus      QInputDevice__DeviceType = 16
	QInputDevice__Airbrush    QInputDevice__DeviceType = 32
	QInputDevice__Keyboard    QInputDevice__DeviceType = 4096
	QInputDevice__AllDevices  QInputDevice__DeviceType = 2147483647
)

type QInputDevice__Capability int

const (
	QInputDevice__None               QInputDevice__Capability = 0
	QInputDevice__Position           QInputDevice__Capability = 1
	QInputDevice__Area               QInputDevice__Capability = 2
	QInputDevice__Pressure           QInputDevice__Capability = 4
	QInputDevice__Velocity           QInputDevice__Capability = 8
	QInputDevice__NormalizedPosition QInputDevice__Capability = 32
	QInputDevice__MouseEmulation     QInputDevice__Capability = 64
	QInputDevice__PixelScroll        QInputDevice__Capability = 128
	QInputDevice__Scroll             QInputDevice__Capability = 256
	QInputDevice__Hover              QInputDevice__Capability = 512
	QInputDevice__Rotation           QInputDevice__Capability = 1024
	QInputDevice__XTilt              QInputDevice__Capability = 2048
	QInputDevice__YTilt              QInputDevice__Capability = 4096
	QInputDevice__TangentialPressure QInputDevice__Capability = 8192
	QInputDevice__ZPosition          QInputDevice__Capability = 16384
	QInputDevice__All                QInputDevice__Capability = 2147483647
)

type QInputDevice struct {
	h          uintptr
	isSubclass bool
}

// NewQInputDevice constructs a new QInputDevice object.
func NewQInputDevice() *QInputDevice {
	ret := newQInputDevice(QInputDevice_new())
	ret.isSubclass = true
	return ret
}

// NewQInputDevice2 constructs a new QInputDevice object.
func NewQInputDevice2(name string, systemId int64, typeVal DeviceType) *QInputDevice {
	name_ms := struct_miqt_string{}
	name_ms.data = CString(name)
	name_ms.len = size_t(len(name))
	defer free(unsafe.Pointer(name_ms.data))

	ret := newQInputDevice(QInputDevice_new2(name_ms, (longlong)(systemId), typeVal))
	ret.isSubclass = true
	return ret
}

// NewQInputDevice3 constructs a new QInputDevice object.
func NewQInputDevice3(parent *QObject) *QInputDevice {
	ret := newQInputDevice(QInputDevice_new3(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQInputDevice4 constructs a new QInputDevice object.
func NewQInputDevice4(name string, systemId int64, typeVal DeviceType, seatName string) *QInputDevice {
	name_ms := struct_miqt_string{}
	name_ms.data = CString(name)
	name_ms.len = size_t(len(name))
	defer free(unsafe.Pointer(name_ms.data))
	seatName_ms := struct_miqt_string{}
	seatName_ms.data = CString(seatName)
	seatName_ms.len = size_t(len(seatName))
	defer free(unsafe.Pointer(seatName_ms.data))

	ret := newQInputDevice(QInputDevice_new4(name_ms, (longlong)(systemId), typeVal, seatName_ms))
	ret.isSubclass = true
	return ret
}

// NewQInputDevice5 constructs a new QInputDevice object.
func NewQInputDevice5(name string, systemId int64, typeVal DeviceType, seatName string, parent *QObject) *QInputDevice {
	name_ms := struct_miqt_string{}
	name_ms.data = CString(name)
	name_ms.len = size_t(len(name))
	defer free(unsafe.Pointer(name_ms.data))
	seatName_ms := struct_miqt_string{}
	seatName_ms.data = CString(seatName)
	seatName_ms.len = size_t(len(seatName))
	defer free(unsafe.Pointer(seatName_ms.data))

	ret := newQInputDevice(QInputDevice_new5(name_ms, (longlong)(systemId), typeVal, seatName_ms, parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QInputDevice) MetaObject() *QMetaObject {
	return newQMetaObject(QInputDevice_MetaObject(this.h))
}

func (this *QInputDevice) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QInputDevice_Metacast(this.h, param1_Cstring))
}

func QInputDevice_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QInputDevice_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QInputDevice) Name() string {
	var _ms struct_miqt_string = QInputDevice_Name(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QInputDevice) Type() DeviceType {
	xxxxxxxxx
}

func (this *QInputDevice) Capabilities() Capabilities {
	xxxxxxxxx
}

func (this *QInputDevice) HasCapability(cap Capability) bool {
	return (bool)(QInputDevice_HasCapability(this.h, cap))
}

func (this *QInputDevice) SystemId() int64 {
	return (int64)(QInputDevice_SystemId(this.h))
}

func (this *QInputDevice) SeatName() string {
	var _ms struct_miqt_string = QInputDevice_SeatName(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QInputDevice) AvailableVirtualGeometry() *QRect {
	_goptr := newQRect(QInputDevice_AvailableVirtualGeometry(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func QInputDevice_SeatNames() []string {
	var _ma struct_miqt_array = QInputDevice_SeatNames()
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret
}

func QInputDevice_Devices() []*QInputDevice {
	var _ma struct_miqt_array = QInputDevice_Devices()
	_ret := make([]*QInputDevice, int(_ma.len))
	_outCast := (*[0xffff]*QInputDevice)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_ret[i] = newQInputDevice(_outCast[i])
	}
	return _ret
}

func QInputDevice_PrimaryKeyboard() *QInputDevice {
	return newQInputDevice(QInputDevice_PrimaryKeyboard())
}

func (this *QInputDevice) OperatorEqual(other *QInputDevice) bool {
	return (bool)(QInputDevice_OperatorEqual(this.h, other.cPointer()))
}

func (this *QInputDevice) AvailableVirtualGeometryChanged(area QRect) {
	QInputDevice_AvailableVirtualGeometryChanged(this.h, area.cPointer())
}

func (this *QInputDevice) OnAvailableVirtualGeometryChanged(slot func(area QRect)) {
	QInputDevice_connect_AvailableVirtualGeometryChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QInputDevice_AvailableVirtualGeometryChanged
func miqt_exec_callback_QInputDevice_AvailableVirtualGeometryChanged(cb intptr_t, area *QRect) {
	gofunc, ok := cgo.Handle(cb).Value().(func(area QRect))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	area_goptr := newQRect(area)
	area_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	slotval1 := *area_goptr

	gofunc(slotval1)
}

func QInputDevice_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QInputDevice_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QInputDevice_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QInputDevice_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QInputDevice_PrimaryKeyboard1(seatName string) *QInputDevice {
	seatName_ms := struct_miqt_string{}
	seatName_ms.data = CString(seatName)
	seatName_ms.len = size_t(len(seatName))
	defer free(unsafe.Pointer(seatName_ms.data))
	return newQInputDevice(QInputDevice_PrimaryKeyboard1(seatName_ms))
}

func (this *QInputDevice) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QInputDevice_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QInputDevice) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QInputDevice_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QInputDevice_MetaObject
func miqt_exec_callback_QInputDevice_MetaObject(self QInputDevice, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QInputDevice{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QInputDevice) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QInputDevice_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QInputDevice) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QInputDevice_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QInputDevice_Metacast
func miqt_exec_callback_QInputDevice_Metacast(self QInputDevice, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QInputDevice{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
